File: Data Structures & Algorithms/count-number-of-islands/submission-4.java
class Solution {
    public int numIslands(char[][] grid) {


        if(grid==null || grid.length==0 || grid[0].length==0 ){
            return 0;
        }

        int count=0;

        for(int i=0;i<grid.length;i++){
            for(int j=0;j<grid[0].length;j++){
                if(grid[i][j]=='1'){
                    count++;
                    bfs(grid, i, j);
                }
            }
        }
        return count;
    }

    public void bfs(char[][] grid, int i, int j){
        Queue<int[]> queue=new LinkedList<>();
        queue.offer(new int[] {i,j});
        grid[i][j]='0';

        while(!queue.isEmpty()){
            int[] cell=queue.poll();
            int row=cell[0];
            int col=cell[1];

            int[][] directions = {{-1,0},{1,0},{0,1},{0,-1}};
            for(int dir[]: directions){
                int newRow=row+dir[0];
                int newCol=col+dir[1];

                if(newRow>=0 && newRow<grid.length && newCol>=0 && newCol<grid[0].length && grid[newRow][newCol]=='1'){
                    grid[newRow][newCol]='0';
                    queue.offer(new int[]{newRow, newCol});
                }
            }
        }

    }
}

File: Data Structures & Algorithms/rotting-fruit/submission-4.py
from collections import deque

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        fresh_count=0
        minutes=0
        directions=[[0,1],[0,-1],[1,0],[-1,0]]
        queue=deque()
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 2:  # rotten
                    queue.append((i,j))
                elif grid[i][j]== 1:  # fresh
                    fresh_count+=1
        
        if fresh_count==0:
            return minutes
        
        while queue:
            for i in range(len(queue)):
                row, col = queue.popleft()
                for dr,dc in directions:
                    new_row=dr+row
                    new_col=dc+col
                
                    if new_row >=0 and new_row<len(grid) and new_col>=0 and new_col<len(grid[0]) and grid[new_row][new_col]==1:
                        grid[new_row][new_col]=2 # mark as rotten
                        fresh_count-=1
                        queue.append((new_row,new_col))
            if queue:
                minutes+=1
        
        if fresh_count == 0:
            return minutes
        else: 
            return -1
        